### PR TITLE
Broken Swagger with Microservices

### LINKED ISSUE #13444: Broken Swagger with Microservices
##### **Overview of the issue**

With current v7 beta, you can't use swagger to call your other microservices

##### **Reproduce the error**

- create a Gateway application
- create a Microservice with 1 entity
- start registry or consul
- start gateway
- start microservice
- try to access to entity with Swagger

##### **Related issues**

Not sure, but I think it's related to upgrade to openapi3

##### **Browsers and Operating System**

-   [ ] Checking this box is mandatory (this is just to show you read everything)

### PR DESCRIPTION
**Pull request by qmonmert**
Fix #13444 

---

Please make sure the below checklist is followed for Pull Requests.

-   [x] [All continuous integration tests](https://github.com/jhipster/generator-jhipster/actions) are green
-   [x] Tests are added where necessary
-   [x] [jhipster-online](https://github.com/jhipster/jhipster-online) was updated if necessary
-   [x] Documentation is added/updated where necessary
-   [x] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/main/CONTRIBUTING.md) are followed

When you are still working on the PR, consider converting it to Draft (bellow reviewers) and adding `skip-ci` label, you can still see CI build result at your branch.

### CODE REVIEW

File: generators/server/templates/src/test/java/package/gateway/responserewriting/SwaggerBasePathRewritingFilterTest.java.ejs [hunk 1]
```diff
@@ -80,13 +80,13 @@ class SwaggerBasePathRewritingFilterTest {
         context.setResponseGZipped(false);
         context.setResponse(response);
 
-        InputStream in = IOUtils.toInputStream("{\"basePath\":\"/\"}", StandardCharsets.UTF_8);
+        InputStream in = IOUtils.toInputStream("{\"servers\":[{\"url\":\"http://localhost:8080/services\"}]}", StandardCharsets.UTF_8);
```
Comment (pascalgrimaud) [inline generators/server/templates/src/test/java/package/gateway/responserewriting/SwaggerBasePathRewritingFilterTest.java.ejs:83]: Line 75 should be instead:

```java
MockHttpServletRequest request = new MockHttpServletRequest("GET", "/services/service1/v3/api-docs");
```

Then, this line should be:

```java
InputStream in = IOUtils.toInputStream("{\"servers\":[{\"url\":\"http://localhost:8080\"}]}", StandardCharsets.UTF_8);

File: generators/server/templates/src/test/java/package/gateway/responserewriting/SwaggerBasePathRewritingFilterTest.java.ejs [hunk 2]
```diff
@@ -99,14 +99,14 @@ class SwaggerBasePathRewritingFilterTest {
         context.setResponseGZipped(true);
         context.setResponse(response);
 
-        context.setResponseDataStream(new ByteArrayInputStream(gzipData("{\"basePath\":\"/\"}")));
+        context.setResponseDataStream(new ByteArrayInputStream(gzipData("{\"servers\":[{\"url\":\"http://localhost:8080/services\"}]}")));
```
Comment (pascalgrimaud) [inline generators/server/templates/src/test/java/package/gateway/responserewriting/SwaggerBasePathRewritingFilterTest.java.ejs:102]: Line 94 should be:

```java
MockHttpServletRequest request = new MockHttpServletRequest("GET", "/services/service1/v3/api-docs");
```

Then, this line should be:

```java
context.setResponseDataStream(new ByteArrayInputStream(gzipData("{\"servers\":[{\"url\":\"http://localhost:8080\"}]}")));
```

File: generators/server/templates/src/test/java/package/gateway/responserewriting/SwaggerBasePathRewritingFilterTest.java.ejs [hunk 3]
```diff
@@ -99,14 +99,14 @@ class SwaggerBasePathRewritingFilterTest {
...
 
         filter.run();
 
         assertEquals("UTF-8", response.getCharacterEncoding());
 
         InputStream responseDataStream = new GZIPInputStream(context.getResponseDataStream());
         String responseBody = IOUtils.toString(responseDataStream, StandardCharsets.UTF_8);
-        assertEquals("{\"basePath\":\"/service1\"}", responseBody);
+        assertEquals("{\"servers\":[{\"url\":\"http://localhost:8080/services/service1\"}]}", responseBody);
     }
```
Comment (pascalgrimaud) [inline generators/server/templates/src/test/java/package/gateway/responserewriting/SwaggerBasePathRewritingFilterTest.java.ejs:111]: add 2 more tests:

Without servers:

```
context.setResponseDataStream(new ByteArrayInputStream(gzipData("{}")));
```

Without URLs:

```
context.setResponseDataStream(new ByteArrayInputStream(gzipData("{\"servers\":[]}")));
```

Then, change your original code according to the tests

File: generators/server/templates/src/main/java/package/gateway/responserewriting/SwaggerBasePathRewritingFilter.java.ejs [hunk 4]
```diff
@@ -95,7 +103,12 @@ public class SwaggerBasePathRewritingFilter extends SendResponseFilter {
                 LinkedHashMap<String, Object> map = this.mapper.readValue(response, LinkedHashMap.class);
 
                 String basePath = requestUri.replace("/v3/api-docs", "");
-                map.put("basePath", basePath);
+                String serversUrl = (String) ((List<Map<String, String>>) map.get(SERVERS)).get(0).get(URL);
```
Comment (pascalgrimaud) [inline generators/server/templates/src/main/java/package/gateway/responserewriting/SwaggerBasePathRewritingFilter.java.ejs:106]: add the new tests, and then change the logic here

**Review by pascalgrimaud — CHANGES_REQUESTED** [review]

**Comment by pascalgrimaud** [comment]
Thanks for the fix @qmonmert !
I added some comments to have better code and coverage

File: generators/server/templates/src/test/java/package/gateway/responserewriting/SwaggerBasePathRewritingFilterTest.java.ejs (hunk 3 above)
Comment (qmonmert) [reply generators/server/templates/src/test/java/package/gateway/responserewriting/SwaggerBasePathRewritingFilterTest.java.ejs:111]: @pascalgrimaud what should be the results in these two cases ?

Comment (qmonmert) [reply generators/server/templates/src/test/java/package/gateway/responserewriting/SwaggerBasePathRewritingFilterTest.java.ejs:111]: Maybe this `assertEquals("{\"servers\":[{\"url\":\"http://localhost:8080/services\"}]}", context.getResponseBody());
`

Comment (pascalgrimaud) [reply generators/server/templates/src/test/java/package/gateway/responserewriting/SwaggerBasePathRewritingFilterTest.java.ejs:111]: the result should be inchanged:

```java
assertEquals("{\"servers\":[{\"url\":\"http://localhost:8080/services/service1\"}]}", responseBody);
```

File: generators/server/templates/src/test/java/package/gateway/responserewriting/SwaggerBasePathRewritingFilterTest.java.ejs (hunk 1 above)
Comment (qmonmert) [reply generators/server/templates/src/test/java/package/gateway/responserewriting/SwaggerBasePathRewritingFilterTest.java.ejs]: ok

File: generators/server/templates/src/test/java/package/gateway/responserewriting/SwaggerBasePathRewritingFilterTest.java.ejs (hunk 3 above)
Comment (qmonmert) [reply generators/server/templates/src/test/java/package/gateway/responserewriting/SwaggerBasePathRewritingFilterTest.java.ejs:111]: ok

File: generators/server/templates/src/main/java/package/gateway/responserewriting/SwaggerBasePathRewritingFilter.java.ejs (hunk 4 above)
Comment (qmonmert) [reply generators/server/templates/src/main/java/package/gateway/responserewriting/SwaggerBasePathRewritingFilter.java.ejs]: ok

**Pull request merged**